recipe/recipe.go

```diff
@@ -36,3 +42,335 @@ type instance struct {
 type file struct {
 	Description string `bson:"description,omitempty" json:"description,omitempty"`
 }
+
+//validateInstance - checks if fields of OutputInstances are not empty for ValidateAddRecipe and ValidateAddInstance
+func (instance *Instance) validateInstance() (missingFields []string, invalidFields []string) {
+
+	if instance.DatasetID == "" {
+		missingFields = append(missingFields, "dataset-id")
+	}
+
+	if instance.Editions != nil && len(instance.Editions) > 0 {
+		for j, edition := range instance.Editions {
+			if edition == "" {
+				missingFields = append(missingFields, "editions["+strconv.Itoa(j)+"]")
+			}
+		}
+	} else {
+		missingFields = append(missingFields, "editions")
+	}
+
+	if instance.Title == "" {
+		missingFields = append(missingFields, "title")
+	}
+
+	if instance.CodeLists != nil && len(instance.CodeLists) > 0 {
+		for i, codelist := range instance.CodeLists {
+			codelistMissingFields, codelistInvalidFields := codelist.validateCodelists()
+
+			if len(codelistMissingFields) > 0 {
+				for _, mField := range codelistMissingFields {
+					codelistMissingFields[i] = "codelists[" + strconv.Itoa(i) + "]." + mField
+				}
+			}
+			if len(codelistInvalidFields) > 0 {
+				for _, iField := range codelistInvalidFields {
+					codelistInvalidFields[i] = "codelists[" + strconv.Itoa(i) + "]." + iField
+				}
+			}
+
+			missingFields = append(missingFields, codelistMissingFields...)
+			invalidFields = append(invalidFields, codelistInvalidFields...)
+		}
+	} else {
+		missingFields = append(missingFields, "codelists")
+	}
+
+	return missingFields, invalidFields
+}
+
+//validateCodelists - checks if fields of CodeList are not empty for ValidateAddRecipe, ValidateAddInstance, ValidateAddCodelist
+func (codelist *CodeList) validateCodelists() (missingFields []string, invalidFields []string) {
+	hrefURL := "http://localhost:22400/code-lists/"
+
+	if codelist.ID == "" {
+		missingFields = append(missingFields, "id")
+	}
+
+	if codelist.HRef == "" {
+		missingFields = append(missingFields, "href")
+	} else {
+		if codelist.HRef != hrefURL+codelist.ID {
```

Comment:
Can we remove or change this check please? I know with the data we have right now it seems like everything must fit that format, but actually we're gonna have to update them all at some point to be more like `localhost:22400/code-lists/ID/editions/edition-id` so it would be good if this validation wasnt preventing us doing that.

Maybe this should just check that its a valid URL? you could try parsing it with the URL package, and making sure that there is a host and a path, and that the path contains '/code-lists' and the id?